enhancements/kube-controller-manager/max-namespace-size.md

```diff
@@ -0,0 +1,112 @@
+---
+title: limit-resources-per-namespace
+authors:
+  - "@deads2k"
+reviewers:
+  - "@soltysh"
+approvers:
+  - "@mfojtik"
+creation-date: 2020-01-31
+last-updated: 2020-01-31
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Limit Resources Per Namespace
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Create a new `namespace-size-policy-controller` in the existing [cluster-policy-controller](https://github.com/openshift/cluster-policy-controller/).
+This controller will create quota objects in every namespace that limit the size of certain error-prone resources.
+We will start with secrets and configmaps.
+It will be created with a goal of protecting the cluster from accidental runaway resource creation.
+
+## Motivation
+
+Kuberentes clusters are vulnerable to runaway namespace conditions.
+This is true regardless of how the cluster is hosted.
+While the true concern is total size of objects in etcd, the immediate cause is usually an excessive number of resources in a single namespace.
+We can easily carve off enforcement of total object count per namespace in a non-disruptive way that protects against common cases of errors.
+
+### Goals
+
+1. protect the cluster from accidental runaway resource creation.
+2. prevent runaway growth instead of simply warning about large sizes
+
+### Non-Goals
+
+1. thwarting evil users trying to break the cluster
+
+## Proposal
+
+Create a resource quota in every namespace that looks roughly like
+```yaml
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  generateName: namespace-size-policy-
+  labels:
+    "clusterpolicycontroller.openshift.io/namespace-size": "true"
+spec:
+    hard:
+      count/secrets: "10000"
+      count/configmaps: "10000"
+```
+There will be rules to ensure that quotas aren't likely to disrupt existing workloads and to allow cluster-admins to effectively bypass the limits.
+
+Rules for how to create namespace-size-policy- quota resources
```

Comment:
How do we deal with existing namesapces that are already out of hand? if there is 100k secrets we will create quota of 10M which likely won't help. Should we mark those quotas that are higher then the default of 10k and not explicitly overridden and trigger alerts on those? Admin could clear the alerts by explicitly overriding the quota.